backend/app/routers/metadata_datasets.py

```diff
@@ -0,0 +1,223 @@
+    context JSON-LD object, context_url, or definition (name of a metadata definition) to be valid.
+
+    Returns:
+        Metadata document that was added to database
+    """
+    if (
+        dataset := await db["datasets"].find_one({"_id": ObjectId(dataset_id)})
+    ) is not None:
+        md = await _build_metadata_db_obj(db, metadata_in, dataset, user)
+        new_metadata = await db["metadata"].insert_one(md.to_mongo())
+        found = await db["metadata"].find_one({"_id": new_metadata.inserted_id})
+        metadata_out = MetadataOut.from_mongo(found)
+        return metadata_out
+
+
+@router.put("/{dataset_id}/metadata", response_model=MetadataOut)
+async def replace_dataset_metadata(
+    metadata_in: MetadataIn,
+    dataset_id: str,
+    user=Depends(get_current_user),
+    db: MongoClient = Depends(dependencies.get_db),
+):
+    """Update metadata. Any fields provided in the contents JSON will be added or updated in the metadata. If context or
+    agent should be changed, use PUT.
+
+    Returns:
+        Metadata document that was updated
+    """
+    if (dataset := await db["datasets"].find_one({"_id": ObjectId(dataset_id)})) is not None:
+        query = {"resource.resource_id": ObjectId(dataset_id)}
+
+        # Filter by MetadataAgent
+        extractor_info = metadata_in.extractor_info
+        if extractor_info is not None:
+            if (
+                    extractor := await db["extractors"].find_one(
+                        {"name": extractor_info.name, "version": extractor_info.version}
+                    )
+            ) is not None:
+                agent = MetadataAgent(creator=user, extractor=extractor)
+                # TODO: How do we handle two different users creating extractor metadata? Currently we ignore user
+                query["agent.extractor.name"] = agent.extractor.name
+                query["agent.extractor.version"] = agent.extractor.version
+            else:
+                raise HTTPException(status_code=404, detail=f"Extractor not found")
+        else:
+            agent = MetadataAgent(creator=user)
+            query["agent.creator.id"] = agent.creator.id
+
+        if (md := await db["metadata"].find_one(query)) is not None:
+            # Metadata exists, so prepare the new document we are going to replace it with
+            md_obj = _build_metadata_db_obj(db, metadata_in, dataset, user, agent=agent)
+            new_metadata = await db["metadata"].replace_one(
+                {"_id": md["_id"]},
+                md_obj.to_mongo())
+            found = await db["metadata"].find_one({"_id": md["_id"]})
+            metadata_out = MetadataOut.from_mongo(found)
+            return metadata_out
+    else:
+        raise HTTPException(status_code=404, detail=f"Dataset {dataset_id} not found")
+
+
+@router.patch("/{dataset_id}/metadata", response_model=MetadataOut)
+async def update_dataset_metadata(
```

Comment:
I'm trying to patch an exist metadata entry by giving the "definition", "metadata id" as well as the "contents". But somehow it's not patching the correct metadata?

My request is
```
{"id":"628664405685a1694705c863","definition":"LatLon","contents":{"latitude":"111","longitude":"222"}}
```

and my response is on another metadata definition:
```
{"id":"6286642c5685a1694705c85f","context":null,"context_url":"clowder.org","definition":null,"contents":{"color":"blue","latitude":"111","longitude":"222"},"resource":{"collection":"datasets","resource_id":"627a8d09ca3d2920a17f602a","version":null},"agent":{"id":"6286642c5685a1694705c860","creator":{"id":"627a8d01ca3d2920a17f6025","email":"cwang138@illinois.edu","first_name":"Chen","last_name":"Wang"},"extractor":null},"created":"2022-05-19T15:37:16.065000"}
```